contrib/native/client/src/clientlib/wincert.ipp

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if defined(IS_SSL_ENABLED)
+
+#include <openssl/x509.h>
+#include <openssl/ssl.h>
+
+#if defined _WIN32  || defined _WIN64
+
+#include <stdio.h>
+#include <windows.h>
+#include <wincrypt.h>
+#include <cryptuiapi.h>
+#include <iostream>
+#include <tchar.h>
+
+
+#pragma comment (lib, "crypt32.lib")
+#pragma comment (lib, "cryptui.lib")
+
+#define MY_ENCODING_TYPE  (PKCS_7_ASN_ENCODING | X509_ASN_ENCODING)
+
+inline
+int loadSystemTrustStore(const SSL *ssl, std::string& msg) {
```

Comment:
maybe it should be a configurable callback where the drill client provides the certificate, and the callback returns if the certificate is valid or not? this way it would be easier to add support for system keystore support on mac or linux too...

Reply:
And that is how boost does it too. But then checking for the certificate is complicated stuff, so it is easier to just load the certificates from the system store into openssl and let openssl deal with it.
Anyway, I think that this counts as an enhancement. I can take care of this when I add support for Mac Keychain.

Reply:
it looks like boost::asio support both loading a file and/or a verify callback:
- http://www.boost.org/doc/libs/1_47_0/doc/html/boost_asio/reference/ssl__context/set_verify_callback/overload1.html
- http://www.boost.org/doc/libs/1_47_0/doc/html/boost_asio/reference/ssl__context/load_verify_file.html

It seems you wouldn't even need to access the native handle when using these functions

Reply:
I'm actually already using these methods (see `SSLStreamChannel::init()`). The verification callback implements validating the certificate. In our case we are using the boost provided rfc2818 verification method. The load verify file should point to the truststore containing the certificates in pem format. OpenSSL will read this file and load the certificate into its in-memory X509 certificate store.
The `loadSystemTrustStore` method reads the certificates from the Windows store (probably the registry) converts from the native store format into X509 and then loads it into the in-memory store. After that OpenSSL takes over and does the verification. 
For Keychain, we will have to do something similar. 
Writing our own certificate verification is going to be error prone, especially if you want to do rfc2818 verification. Not sure I'm up to it :(.